src/dune_pkg/opam.ml

```diff
@@ -0,0 +1,188 @@
+open Stdune
+module Package_name = Dune_lang.Package_name
+
+module Repo = struct
+  let ( / ) = Filename.concat
+
+  type t = { packages_dir_path : string }
+
+  let of_opam_repo_dir_path opam_repo_dir_path =
+    if not (Sys.file_exists opam_repo_dir_path) then
+      User_error.raise [ Pp.textf "\"%s\" does not exist" opam_repo_dir_path ];
+    if not (Sys.is_directory opam_repo_dir_path) then
+      User_error.raise
+        [ Pp.textf "\"%s\" is not a directory" opam_repo_dir_path ];
+    let packages_dir_path = opam_repo_dir_path / "packages" in
+    if
+      not
+        (Sys.file_exists packages_dir_path && Sys.is_directory packages_dir_path)
```

Comment:
I think this should check for the `repo` file too, which as far as I am aware needs to contin at least `opam-version: "2.0"`.

Reply:
Is it not sufficient to let opam check the `opam-version` in the `opam` files in the repo as it's parsing them? Or would this just be to improve UX in that we could catch the problem earlier and print a more meaningful error message than the error opam would produce while parsing files? I left this file out as it's not necessary to solve dependencies but it would be easy to add. Do you happen to know if the opam libraries provide an easy way to parse this file?

Reply:
I would say we should require the same format for the repo as OPAM to not introduce dialects ("this repo only works with dune"/"this repo only works with OPAM"). If I see correctly, `opam-format` has `OpamFile.Repo` which defines accessors to the metadata that can be found in the `repo` file so I would say as long as it can parse it it's a valid repo.